File: Dynamic programming/leetcode_05.cpp

// Problem Link ====>> https://leetcode.com/problems/longest-palindromic-substring/description/

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    string longestPalindrome(string s)
    {
        int n = s.length();
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        int maxLen = 1;
        int start = 0;
        for (int i = 0; i < n; i++)
        {
            dp[i][i] = true;
        }

        for (int i = 0; i < n - 1; i++)
        {
            if (s[i] == s[i + 1])
            {
                dp[i][i + 1] = true;
                maxLen = 2;
                start = i;
            }
        }

        for (int len = 3; len <= n; len++)
        {
            for (int i = 0; i + len - 1 < n; i++)
            {
                if (s[i] == s[i + len - 1] && dp[i + 1][i + len - 2])
                {
                    dp[i][i + len - 1] = true;
                    if (len > maxLen)
                    {
                        maxLen = len;
                        start = i;
                    }
                }
            }
        }

        return s.substr(start, maxLen);
    }
};

int main()
{
    string s = "babad";
    Solution solution;
    string result = solution.longestPalindrome(s);
    cout << result << endl;

    return 0;
}

File: Dynamic programming/leetcode_05.py

# Problem Link ====>> https://leetcode.com/problems/longest-palindromic-substring/description/

class Solution:
  def longestPalindrome(self, s):
    n = len(s)
    dp = [[False for j in range(n)] for i in range(n)]
    maxLen = 1
    start = 0
    for i in range(n):
      dp[i][i] = True

    for i in range(n - 1):
      if s[i] == s[i + 1]:
        dp[i][i + 1] = True
        maxLen = 2
        start = i

    for len in range(3, n + 1):
      for i in range(n - len + 1):
        if s[i] == s[i + len - 1] and dp[i + 1][i + len - 2]:
          dp[i][i + len - 1] = True
          if len > maxLen:
            maxLen = len
            start = i

    return s[start: start + maxLen]

def main():
  s = "babad"
  solution = Solution()
  result = solution.longestPalindrome(s)
  print(result)

if __name__ == "__main__":
  main()
